### PR TITLE
feat(md): versions history page

### PR DESCRIPTION
**Pull request by ranihorev**
This PR introduces the history page back. 
I had to reorganize some existing components in order to make them reusable. 
The nice part is that I was able to use a lot of the existing components :)

### CODE REVIEW

File: app/scripts/modules/core/src/managed/versionsHistory/VersionContent.tsx [hunk 1]
```diff
@@ -14,7 +20,18 @@ interface IVersionInEnvironmentProps {
...
 }
 
+const useGetDetailedVersionData = ({ environment, version }: Omit<IVersionInEnvironmentProps, 'envPinnedVersions'>) => {
+  const app = useApplicationContextSafe();
+  const { data: detailedVersionData, error, loading } = useFetchVersionQuery({
+    variables: { appName: app.name, versions: [version.version] },
+  });
+  const environmentData = detailedVersionData?.application?.environments?.find((env) => env.name === environment);
+  const artifactData = environmentData?.state.artifacts?.find((artifact) => artifact.reference === version.reference);
+  return { detailedVersionData: artifactData?.versions?.find((v) => v.version === version.version), error, loading };
```
Comment (christopherthielen) [inline app/scripts/modules/core/src/managed/versionsHistory/VersionContent.tsx:30]: There are lots of optional fields in these interfaces.  Can you comment about why that is?  Is this an artifact of the gql schema?

File: app/scripts/modules/core/src/managed/versionsHistory/VersionContent.tsx [hunk 2]
```diff
@@ -40,13 +58,26 @@ const VersionInEnvironment = ({ environment, version, envPinnedVersions }: IVers
...
+        deployedAt={detailedVersionData?.deployedAt}
+        buildDuration={getLifecycleEventDuration(detailedVersionData, 'BUILD')}
         actions={actions}
         isPinned={isPinned}
       />
+      {loading && (
+        <ContentLoader width="100%" height={30}>
+          <rect x="0" y="8" rx="5" ry="5" width="60%" height="8" />
+          <rect x="0" y="22" rx="5" ry="5" width="50%" height="8" />
+        </ContentLoader>
```
Comment (christopherthielen) [inline app/scripts/modules/core/src/managed/versionsHistory/VersionContent.tsx]: should/could this be pulled out to a separate component?

File: app/scripts/modules/core/src/managed/versionsHistory/VersionHeading.tsx [hunk 3]
```diff
@@ -57,8 +61,16 @@ interface IVersionHeadingProps {
...
 export const VersionHeading = ({ group, chevron }: IVersionHeadingProps) => {
   const gitMetadata = group.gitMetadata;
+  const client = useApolloClient();
+  const app = useApplicationContextSafe();
+  const prefetchData = () => {
+    client.query<any, FetchVersionQueryVariables>({
+      query: FetchVersionDocument,
+      variables: { appName: app.name, versions: Array.from(group.versions) },
+    });
+  };
```
Comment (christopherthielen) [inline app/scripts/modules/core/src/managed/versionsHistory/VersionHeading.tsx:82]: is it worth commenting where the results of this query is expected to be consumed?

File: app/scripts/modules/core/src/managed/versionsHistory/VersionContent.tsx [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import React from 'react';
-import ContentLoader from 'react-content-loader';
+// import ContentLoader from 'react-content-loader';
```
Comment (christopherthielen) [inline app/scripts/modules/core/src/managed/versionsHistory/VersionContent.tsx:2]: as a general note, we usually discuss with the group before adding new deps
this one seems teeny though, so I'm 👍

**Review by christopherthielen — APPROVED** [review]
Really great work.

❤️ Split into reviewable commits
❤️ Great work breaking down components into smaller ones
❤️ Love your diligence on great TS types

File: app/scripts/modules/core/src/managed/environmentBaseElements/BaseEnvironmentItem.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+interface IBaseEnvironmentItemProps {
+  title: string | React.ReactElement;
+  className?: string;
+  iconTooltip: string;
+  iconName: IIconProps['name'];
+  size?: 'regular' | 'small';
+}
+
+export const BaseEnvironmentItem: React.FC<IBaseEnvironmentItemProps> = ({
```
[the remark this answers is not in the document]
Comment (ranihorev) [reply app/scripts/modules/core/src/managed/environmentBaseElements/BaseEnvironmentItem.tsx:17]: Yeah, `EnvironmentItem` is probably better

File: app/scripts/modules/core/src/managed/graphql/graphql-sdk.ts [hunk 6]
```diff
@@ -391,6 +391,153 @@ export type FetchApplicationQuery = { __typename?: 'Query' } & {
...
+      }
+  >;
+};
+
+export type FetchVersionQueryVariables = Exact<{
+  appName: Scalars['String'];
+  versions?: Maybe<Array<Scalars['String']> | Scalars['String']>;
+}>;
+
+export type FetchVersionQuery = { __typename?: 'Query' } & {
```
[the remark this answers is not in the document]
Comment (ranihorev) [reply app/scripts/modules/core/src/managed/graphql/graphql-sdk.ts:456]: Why? :)

[the remark this answers is not in the document]
Comment (ranihorev) [reply app/scripts/modules/core/src/managed/graphql/graphql-sdk.ts:456]: This is auto-generated btw

File: app/scripts/modules/core/src/managed/graphql/graphql-sdk.ts [hunk 7]
```diff
@@ -579,6 +726,210 @@ export function useFetchApplicationLazyQuery(
...
+    options,
+  );
+}
+export type FetchVersionsHistoryQueryHookResult = ReturnType<typeof useFetchVersionsHistoryQuery>;
+export type FetchVersionsHistoryLazyQueryHookResult = ReturnType<typeof useFetchVersionsHistoryLazyQuery>;
+export type FetchVersionsHistoryQueryResult = Apollo.QueryResult<
+  FetchVersionsHistoryQuery,
+  FetchVersionsHistoryQueryVariables
+>;
+export const FetchVersionDocument = gql`
```
[the remark this answers is not in the document]
Comment (ranihorev) [reply app/scripts/modules/core/src/managed/graphql/graphql-sdk.ts:822]: Yes, there is a `graphql:generate` command for that. At some point, we can consider automating some of it, but I don't see a need at the moment

File: app/scripts/modules/core/src/managed/graphql/graphql-sdk.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (christopherthielen) [reply app/scripts/modules/core/src/managed/graphql/graphql-sdk.ts:456]: > This is auto-generated btw

yeah, I gathered that, lol 😄   It's just a little cray-cray.

In spotlight, we are generating these files on the fly and only committing the graphql schema to source control.

File: app/scripts/modules/core/src/managed/versionsHistory/VersionContent.tsx (hunk 1 above)
Comment (ranihorev) [reply app/scripts/modules/core/src/managed/versionsHistory/VersionContent.tsx:30]: Yes, in general, it's recommended to use nullable fields in graphql except for identifier, it makes your JS code a bit though... 
I'll share with you a doc with some context

File: app/scripts/modules/core/src/managed/versionsHistory/VersionContent.tsx (hunk 4 above)
Comment (ranihorev) [reply app/scripts/modules/core/src/managed/versionsHistory/VersionContent.tsx:2]: Good to know, thanks

File: app/scripts/modules/core/src/managed/environmentBaseElements/BaseEnvironmentItem.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (ranihorev) [reply app/scripts/modules/core/src/managed/environmentBaseElements/BaseEnvironmentItem.tsx:17]: I'll probably rename many of these components as part of removing the old UI

File: app/scripts/modules/core/src/managed/graphql/graphql-sdk.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ranihorev) [reply app/scripts/modules/core/src/managed/graphql/graphql-sdk.ts:456]: I don't think that it's a good idea tbh:
- How do you use the types locally?
- How do you make sure the local types are identical to the auto-generated types?

[the remark this answers is not in the document]
Comment (christopherthielen) [reply app/scripts/modules/core/src/managed/graphql/graphql-sdk.ts:456]: They get auto-generated on the fly by tooling I guess.  @alanmquach  would know

**Pull request merged**